app/guided.py

```diff
@@ -0,0 +1,424 @@
+            if len(planner_goals) < current_goal_count:
+                absolute_distance_table, effective_distance_table = await self._handle_goal_achieved(attack_graph, goals=planner_goals)
+            else:
+                effective_distance_table = await self._handle_no_goal_achieved(absolute_distance_table,
+                                                                                effective_distance_table,
+                                                                                goal_links)
+            goal_links = await self._get_goal_links(absolute_distance_table, agent=agent)
+
+        if not len(goal_links) or not len(planner_goals):
+            self.planning_svc.log.debug('No more links available or all goals satisfied, planner complete.')
+            self.next_bucket = None
+
+    """ PRIVATE """
+
+    async def _build_attack_graph(self, abilities, agent):
+        """
+        Produces a directed graph that includes nodes for each ability, and nodes of each input and output fact
+        of those abilities.
+        """
+        attack_graph = nx.DiGraph()
+        for ability_id in abilities:
+            for ability in await self.data_svc.locate('abilities', dict(ability_id=ability_id)):
+                if agent:
+                    attack_graph = await self._output_fact_edges(ability, attack_graph, agent)
+                    attack_graph = await self._input_fact_edges(ability, attack_graph, agent)
+                else:
+                    for operation_agent in self.operation.agents:
+                        attack_graph = await self._output_fact_edges(ability, attack_graph, operation_agent)
+                        attack_graph = await self._input_fact_edges(ability, attack_graph, operation_agent)
+        return attack_graph
+
+    async def _output_fact_edges(self, ability, graph, agent):
+        """
+        Adds output facts to a directed graph based on the parserconfigs included in an ability.
+        """
+        executor = await agent.get_preferred_executor(ability)
+        for parser in executor.parsers:
+            for parserconfig in parser.parserconfigs:
+                graph.add_edge(ability, parserconfig.source)
+                if parserconfig.target:
+                    graph.add_edge(ability, parserconfig.target)
+                if parserconfig.edge:
+                    graph.add_edge(ability, parserconfig.edge)
+        return graph
+
+    async def _input_fact_edges(self, ability, graph, agent):
+        """
+        Adds input facts to a directed graph based on the facts present in a command template.
+        """
+        def add_to_graph(fact):
+            if fact not in list(Agent.RESERVED):
+                graph.add_edge(fact, ability)
+
+        executor = await agent.get_preferred_executor(ability)
+        for fact in re.findall(NONLIMITED_FACT_REGEX, executor.test, flags=re.DOTALL):
+            nonlimited = re.search(LIMITED_FACT_REGEX, fact)
```

Comment:
My understanding is that in this line `nonlimited` capturers facts with some limiting condition, such that the limiting condition can be removed before adding the fact to the attack graph with `add_to_graph`.  This is the only case of a variable name which might not be completely self-documenting.